pytorch_lightning/core/lightning.py

```diff
@@ -360,6 +369,17 @@ def log(
             # reset any tensors for the new hook name
             results.reset(metrics=False, fx=self._current_fx_name)
 
+        if attribute_name is None and isinstance(value, Metric):
+            # this is used to effiently find the attribute prefix path of metric objects
+            # this will enable Lightning to re-attach metric reference when reloading states.
```

Comment:
Hey @awaelchli, Thanks for pointing this out. Let me make the description more clear.

Here is the explanation:

The Metrics state_dict get serialised with the model state_dict and the ResultCollection within the loop state dict and doesn't save anything related to TM Metric as it would be duplicated information.
When reloading, the ResultCollection ResultMetric value attribute is None as no metrics have been attached yet.

To re-attach, there are 2 options:

1. Wait for next `self.log` call where the metric would be provided and we gain batch the reference.
Cons: This won't work if the user have conditional logging

2. Save an extra field called attribute_name within ResultMetric Metadata object. The attribute_name is the prefix to find the Metric reference from the LightningModule on reload automatically.

On reload, we recursively explore the LightningModule for Metrics and create a mapping {"attribute_name": Metric}

https://github.com/PyTorchLightning/pytorch-lightning/pull/7966/files#diff-9b4f83dd619c37d176c0ae5d9d460db3ae0f6ebd60cf0577c5393f4e601413d8R260

```
results.load_state_dict(state_dict, sync_fn=sync_fn, metrics=metrics)
```

```py
if metrics:
        for attribute_name, metric in metrics.items():
            for result_metric in self.result_metrics:
                if result_metric.meta.attribute_name == attribute_name:
                    result_metric.value = metric
```

https://github.com/PyTorchLightning/pytorch-lightning/pull/7966/files#diff-2d9e4d82fd49872d44f822eef3125a2b18efa2b32af9c94bff479a267f553be4R678

Reply:
Any idea for a better name or better solution ?

Reply:
Changed to `metric_prefix_name`.